librosa/segment.py

```diff
@@ -50,6 +52,193 @@
+        Cross-similarity matrix
+
+    See Also
+    --------
+    recurrence_matrix
+    recurrence_to_lag
+    feature.stack_memory
+    sklearn.neighbors.NearestNeighbors
+    scipy.spatial.distance.cdist
+
+    Notes
+    -----
+    This function caches at level 30.
+
+    Examples
+    --------
+    Find nearest neighbors in MFCC space between two sequences
+
+    >>> y_ref, sr = librosa.load(librosa.util.example_audio_file())
+    >>> y_comp, sr = librosa.load(librosa.util.example_audio_file(), offset=10)
+    >>> mfcc_ref = librosa.feature.mfcc(y=y_ref, sr=sr)
+    >>> mfcc_comp = librosa.feature.mfcc(y=y_comp, sr=sr)
+    >>> xsim = librosa.segment.cross_similarity(mfcc_ref, mfcc_comp)
+
+    Or fix the number of nearest neighbors to 5
+
+    >>> xsim = librosa.segment.cross_similarity(mfcc_ref, mfcc_comp, k=5)
+
+    Suppress neighbors within +- 7 samples
+
+    >>> xsim = librosa.segment.cross_similarity(mfcc_ref, mfcc_comp, width=7)
+
+    Use cosine similarity instead of Euclidean distance
+
+    >>> xsim = librosa.segment.cross_similarity(mfcc_ref, mfcc_comp, metric='cosine')
+
+    Use an affinity matrix instead of binary connectivity
+
+    >>> xsim_aff = librosa.segment.cross_similarity(mfcc_ref, mfcc_comp, mode='affinity')
+
+    Plot the feature and recurrence matrices
+
+    >>> import matplotlib.pyplot as plt
+    >>> plt.figure(figsize=(8, 4))
+    >>> plt.subplot(1, 2, 1)
+    >>> librosa.display.specshow(xsim, x_axis='time', y_axis='time')
+    >>> plt.title('Binary recurrence (symmetric)')
+    >>> plt.subplot(1, 2, 2)
+    >>> librosa.display.specshow(xsim_aff, x_axis='time', y_axis='time',
+    ...                          cmap='magma_r')
+    >>> plt.title('Affinity recurrence')
+    >>> plt.tight_layout()
+
+    '''
+    data_from = np.atleast_2d(data_from)
+    data_to = np.atleast_2d(data_to)
+
+    if data_from.shape[0] != data_to.shape[0]:
+        raise ValueError("data_from and data_to must have the same first dimension")
+
+    # swap data axes so the feature axis is last
+    data_from = np.swapaxes(data_from, -1, 0)
+    n_from = data_from.shape[0]
+    data_from = data_from.reshape((n_from, -1))
+    data_to = np.swapaxes(data_to, -1, 0)
+    n_to = data_to.shape[0]
+    data_to = data_to.reshape((n_to, -1))
+
+    n_feat = data_from.shape[-1]
+
+    if mode not in ['connectivity', 'distance', 'affinity']:
+        raise ParameterError(("Invalid mode='{}'. Must be one of "
+                              "['connectivity', 'distance', "
+                              "'affinity']").format(mode))
+    if k is None:
+        if n_from > 3:
+            k = 2 * np.ceil(np.sqrt(n_to - 1))
+        else:
+            k = 2
+
+    if bandwidth is not None:
+        if bandwidth <= 0:
+            raise ParameterError('Invalid bandwidth={}. '
+                                 'Must be strictly positive.'.format(bandwidth))
+
+    k = int(k)
+
+    # Build the neighbor search object
+    try:
```

Comment:
can you put a comment explaining why this try-catch statement is necessary? what could go wrong with `algorithm='auto'`?